air/src/air/transition/frame.rs

```diff
@@ -3,80 +3,81 @@
 // This source code is licensed under the MIT license found in the
 // LICENSE file in the root directory of this source tree.
 
-use super::{FieldElement, Vec};
+use super::{super::Air, FieldElement, Vec};
+use utils::iterators::StreamingIterator;
 
-// EVALUATION FRAME
-// ================================================================================================
 /// A set of execution trace rows required for evaluation of transition constraints.
-///
-/// In the current implementation, an evaluation frame always contains two consecutive rows of the
-/// execution trace. It is passed in as one of the parameters into
+/// It is passed in as one of the parameters into
 /// [Air::evaluate_transition()](crate::Air::evaluate_transition) function.
+pub trait EvaluationFrame<E: FieldElement> {
```

Comment:
I think right now we assume that the frame rows are contiguous. Maybe there is a relatively easy way to relax this condition. For example, we could add the following function to the trait:
```Rust
fn offsets() -> &'static [usize];
```

And then for a default implementation this function could return a reference to the following constant:
```Rust
const FRAME_OFFSETS: [usize; 2] = [0, 1];
```

And custom implementations could return different arrays. For example:
```Rust
const FRAME_OFFSETS: [usize; 2] = [0, 1, 3, 31];
```
Would mean that the frame consists of 4 rows: current, next, 3 rows forward, and 31 rows forward.

These are just initial thoughts. I haven't really considered all pros and cons, and there might be a better way of doing this.

Reply:
btw, this would also provide a default implementation of `row_count()` function. It could look like:
```Rust
fn row_count(&self) -> usize {
    Self::offsets().len()
}
```

A separate nit: would prefer `num_rows()` instead of `row_count()` (that's the convention I've been trying to use throughout the library).

Reply:
Good idea. I've also added `FRAME_SHIFT` to indicate how many steps in the trace the frame should skip over at each evaluation. It might be a good idea to indicate not just the rows that need to be read but also columns (I mentioned this in the original issue), but I think this is a good start.

Reply:
Looks good! Two small comments:

* I don't think we can have `FRAME_OFFSETS` as a constant at the trait level as it limits the number of items in the array. Instead, we can make `offsets()` function be a required so that a specific implementation of the trait can adapt it as needed.
* We probably don't need 'shift()' function as `FRAME_SHIFT` constant can be just as easily accessed.